Documentation Update (#210)
mezo-org/musd: docs/README.md

2. MUSD is minted to the PCV contract.
3. MUSD is then deposited from the PCV contract into the Stability Pool.

When the protocol's MUSD is withdrawn from the Stability Pool to the PCV contract it is first used to repay any outstanding protocol loan balance. This is to ensure that the bootstrap loan can not be withdrawn from the protocol.

It is anticipated that repayments on the protocol bootstrap loan will be made on a roughly weekly basis via calls to distributeMUSD in the PCV contract.

### Protocol Controlled Value

Protocol Owned Liquidty as the name suggests is owned by the protocol, it is intended to provide utility to the protocol regardless of market conditions and without requiring incentives for the liquidity to stay in place.

Over time as the protocol accrues interest and fees the bootstrap loan gets repaid and the portion of the MUSD in the Stability Pool that is Protocol Owned Liquidity increases.

#### Distributing fees with an active Protocol Loan

When the protocol is initialised with the fee recipient and an active protocol loan the following occurs.

![Active Protocol Loan and Fees](images/feesAndActiveProtocolLoan.png)

Flow of Funds

> 1. A user transaction triggers an event that results in a fee or interest collection events

Reviewer: Fixed in 6e1ddf427bc195f8fb8ddb1e60dbf3ca34c3fb1d, resolving.

2. MUSD is minted and sent to the PCV contract
3. A governance action triggers the movement of MUSD from the PCV contract to the PCV fee recipient address
4. A portion of the accrued MUSD is sent to the fee recipient
5. A portion of the MUSD is burnt to reduce the protocol loan.

#### Distributing fees when the Protocol Loan has been repaid

When the protocol is initialised with a fee recipient and the protocol loan has been repaid then the following occurs.

![Repaid Protocol Loan and Fees](images/feesAndRepaidProtocolLoan.png)

Flow of Funds

> 1. A user transaction triggers an event that results in a fee or interest collection events

Reviewer: Fixed in 6e1ddf427bc195f8fb8ddb1e60dbf3ca34c3fb1d, resolving.

#### Liquidation when the Stability Pool is empty.

If the Stability Pool is empty, we redistribute both the debt and collateral. All of the debt and collateral is sent to the Default Pool, where a user's ownership of the default pool is equal to their proportional ownership of all deposited collateral. The newly acquired collateral and debt are included for all purposes: calculating collateralization ratio, redemptions, closing a trove, etc.

![Liquidation when the Stability Pool is empty flow](images/liquidationWithEmptyStabilityPool.png)

Flow of Funds

1. Liquidator sends liquidation request to Trove Manager.
2. Liquidator receives gas compensation (MUSD).
3. Liquidator receives 0.5% of the liquidated BTC.
4. BTC is sent from the ActivePool to the DefaultPool for proportionate redistribution to other loans.
5. When active loans next have a borrower operation, redemption or liquidation they take on a proportionate amount of the liquidated debt and collateral which results in their share of the liquidated BTC being moved back to the Active Pool.

> The pending debt does not accrue interest until it has been applied to a borrower's loan.

Reviewer 2: In step 5 it refers to borrower operations, redemptions or liquidations.  We could expand borrower operations to spell them all out if you think that's needed.  I just added a little bit of clarification here: [1359dae](https://github.com/mezo-org/musd/pull/210/commits/1359dae41a7e6e9e18b93cce0878e8bb5d12a356)

Note that when a redistribution of debt and collateral from a liquidated loan is done across the other loans, no external liquidity source is required.

#### Liquidations and Protocol Owned Liquidity

The Stability Pool is initially seeded with a [bootstrap loan](#protocol-bootstrap-loan) against the protocol's future fees, over time the MUSD minted as part of this loan is converted into Protocol Owned Liquidity through fees and interest being burnt by the PCV contract.

BTC acquired by the protocol through liquidations may be converted back into MUSD and redeposited into the Stablity Pool depending on the maturity and state of the protocol at the time of liquidation.

##### Rebalancing Large Liquidations

In the early days of Liquity v1 the protocol almost had to deal with a stress test from a large deposit that was a disproportionate amount of the TVL getting liquiditated. This could have resulted in cascading liquidations with sustained negative price action.

The protocol loan is intended to help mitigate this circumstance.

> If this scenario was to arise early into the life of MUSD where a $100M liquidation emptied the Stability Pool, this might leave $109M of BTC in the Stability Pool and insufficient BTC in the Active Pool to honour redemptions for all of the circulating MUSD. In this event the governance process that manages the PCV contract would withdraw the BTC from the Stability Pool and open a loan to borrow MUSD against BTC to get the BTC back into the Active Pool so that redemptions are able to honour requests. The borrowed MUSD would then be deposited into the Stability Pool.

Reviewer 2: I feel like we can leave them as-is since the code *does* in fact create a 100M bootstrap loan, we just modified it after deployment.  If needed, we could add a mention of this in the docs but I feel like that might confuse things.  What do you think?

### Redemptions

A user may call `TroveManager.redeemCollateral` to burn MUSD to obtain BTC, $1 for $1 worth (minus the redemption fee). This is the main mechanism [maintaining the peg](#maintaining-the-peg).

The trove with the lowest collateral ratio (but above the 110% liquidation threshold) has an equivalent amount of debt canceled, and then their BTC is trasferred to the redeeming user. This has a net effect of _raising_ their collateral ratio.

For example, say that...

- Alice has $1000 debt backed by $1300 collateral (130% ratio).
- Bob has $1000 debt backed by $2000 collatearl (200% ratio).

Carol redeems $50. Alice's debt and collateral are reduced by $50. Carol receives `$50 * .995 = $49.75` worth of BTC, and the protocol receives `$50 * .005 = $0.25` as a redemption fee.

Alice now has $950 debt backed by $1250 collateral (132% ratio).

Someone's full debt can be cancelled in this way. For example, if Carol redeemed $1000 instead of $50, then Alice's debt would be fully paid, and she would be left with $300 worth of collateral. The remaining collateral is sent to the `CollSurplusPool`. Alice can collect it by calling `BorrowerOperations.claimCollateral`.

> The `minNetDebt` value is used to prevent the list of active loans being filled up with nearly empty troves, filling the sorted loans list with lots of nearly empty troves could be used to cause redemption requests to fail by running out of gas. Note that the redemption amount must leave the loan with a debt level greater than the `minNetDebt` value, this is initialized at 1800 MUSD but can be changed through governance, it can be increased or reduced as low as 50 MUSD.

Reviewer 2: I removed the mention of changing the `minNetDebt` since that would have other implications that I think are probably outside the scope of this README: [883b334](https://github.com/mezo-org/musd/pull/210/commits/883b334192b56a6e2bd70866dcf02516d385af2e)

#### Partial Loan Redemption

When the redemption amount is less than the debt of the loan with the lowest collateralisation this results in a partial redemption.

![Partial Loan Redemption](images/partialRedemption.png)

Flow of Funds

1. Redeemer deposits MUSD into Borrower Operation contract.
2. MUSD is burnt.
3. BTC is sent to the Redeemer.
4. Redemption fee is sent to PCV.

#### Full Loan Redemption

When the redemption amount is greater than a loans value, that loan become fully redeemed. The remaining redemption amount is then applied to the next loan, this continues until the full redemption amount is fulfilled.

![Full loan Redemption](images/fullRedemption.png)

Flow of Funds

1. Redeemer deposits MUSD into BorrowerOperation contract and fully redeems a borrowers debt.
2. MUSD is burnt.
3. BTC is sent to the Redeemer.
4. Redemption fee is sent to PCV.
5. Borrowers excess BTC is sent to the CollSurplusPool.

> 6. Liquidation reserve is burnt.

Reviewer 2: Clarified gas compensation here: [64e0ce0](https://github.com/mezo-org/musd/pull/210/commits/64e0ce0c44cfae0f351e050c1c8f60ec25b3a2ca)

Liquidations can be triggered on individual loans with `TroveManager.liquidate` or on a list of loans with `TroveManager.batchLiquidateTroves`.

#### Liquidation using the Stability Pool

When the Stability Pool has sufficent funds to cover the liquidated loan/s debt, the Stability Pool burns MUSD to cover all of the trove's debt and seizes the remaining 99.5% of the trove's collateral. The MUSD burnt to cover the debt is done proportionately across all MUSD deposits in the Stability Pool, in exchange the Stability Pool depositor has a right to claim the same proportion of the collateral from the liquidation.

![Liquidation using the Stability Pool flow](images/liquidationUsingStabilityPool.png)

Flow of Funds

1. Liquidator sends liquidation request to Trove Manager
2. Liquidator receives gas compensation (MUSD)
3. Stability Pool burns the MUSD required to buy the debt
4. Liquidator receives 0.5% of the liquidated BTC
5. Remaining BTC collateral is sent from the Active Pool to the Stability Pool

> Anyone can deposit into the Stabily Pool and proportionately acquire discounted BTC when a liquidation occurs, however it is not expected that users will deposit into the Stability Pool. Unlike Liquity v1 there are no direct incentives for depositing into the Stability Pool. This means the yield from being a Stability Pool depositor would be purely based on the liquidations that occur.

Reviewer 2: ```suggestion
Anyone can deposit into the Stability Pool and proportionately acquire discounted BTC when a liquidation occurs, however it is not expected that users will deposit into the Stability Pool. Unlike Liquity v1 there are no direct incentives for depositing into the Stability Pool. This means the yield from being a Stability Pool depositor would be purely based on the liquidations that occur.
```

Reviewer: Fixed in 6e1ddf427bc195f8fb8ddb1e60dbf3ca34c3fb1d, resolving.

The liquidity source for the Stability Pool is intended to initially be from the [protocol loan](#protocol-bootstrap-loan) and for this balance to grow over time through interest and fees.

#### Liquidation greater than the Stability Pool balance

If there are insufficent funds in the Stability Pool to cover the liquidated loan/s debt, the Stability Pool covers as much debt as it can, burning MUSD in exchange for the collatearl. The remaining debt and collateral is proportionately redistrubuted across the remaining loans.

![Liquidation partially using the Stability Pool flow](images/liquidationPartiallyUsingStabilityPool.png)

Flow of Funds

1. Liquidator sends liquidation request to Trove Manager
2. Liquidator receives gas compensation (MUSD)
3. Stability Pool burns the MUSD required to buy the debt
4. Liquidator receives 0.5% of the liquidated BTC
5. BTC is sent from the Active Pool to the Stability Pool

> 6. BTC for debt that cant be covered by the Stability Pool is sent to the Default Pool

Reviewer: Fixed in 6e1ddf427bc195f8fb8ddb1e60dbf3ca34c3fb1d, resolving.

#### Liquidation when the Stability Pool is empty.

If the Stability Pool is empty, we redistribute both the debt and collateral. All of the debt and collateral is sent to the Default Pool, where a user's ownership of the default pool is equal to their proportional ownership of all deposited collateral. The newly acquired collateral and debt are included for all purposes: calculating collateralization ratio, redemptions, closing a trove, etc.

![Liquidation when the Stability Pool is empty flow](images/liquidationWithEmptyStabilityPool.png)

Flow of Funds

1. Liquidator sends liquidation request to Trove Manager
2. Liquidator receives gas compensation (MUSD)
3. Liquidator receives 0.5% of the liquidated BTC
4. BTC is sent from the ActivePool to the DefaultPool for proportionate redistribution to other loans.
5. When active loans next have a borrower operation, redemption or liquidation they take on a proportionate amount of the liquidated debt and collateral which results in their share of the liquidated BTC being moved back to the Active Pool.

> The pending debt does not accrue interest until it has been applied to a borrowers loan.

Reviewer 2: ```suggestion
The pending debt does not accrue interest until it has been applied to a borrower's loan.
```

Reviewer: Fixed in 6e1ddf427bc195f8fb8ddb1e60dbf3ca34c3fb1d, resolving.

> Note that when a redistribution of debt and collateral from a liquidated loan is done across the other loans, no external liquidty source is required.

Reviewer 2: ```suggestion
Note that when a redistribution of debt and collateral from a liquidated loan is done across the other loans, no external liquidity source is required.
```

Reviewer: Fixed in 6e1ddf427bc195f8fb8ddb1e60dbf3ca34c3fb1d, resolving.

#### Liquidations and Protocol Owned Liquidity

The Stability Pool is initially seeded with a [bootstrap loan](#protocol-bootstrap-loan) against the protocol's future fees, over time the MUSD minted as part of this loan is converted into Protocol Owned Liquidity through fees and interest being burnt by the PCV contract.

BTC acquired by the protocol through liquidations may be converted back into MUSD and redeposited into the Stablity Pool depending on the maturity and state of the protocol at the time of liquidation.

##### Rebalancing Large Liquidations

In the early days of Liquity v1 the protocol almost had to deal with a stress test from a large deposit that was a disproportionate amount of the TVL getting liquiditated. This could have resulted in cascading liquidations with sustained negative price action.

The protocol loan is intended to help mitigate this circumstance.

> If this scenario was to arise early into the life of MUSD where a $100M liquidation emptied the Stability Pool. This might leave $109M of BTC in the Stability Pool and insufficent BTC in the Active Pool to honour redemptions for all of the circulating MUSD. In this event the governance process that manages the PCV contract would withdraw the BTC from the Stability Pool and open a loan to borrow MUSD against BTC to get the BTC back into the Active Pool so that redemptions are able to honour requests. The borrowed MUSD would then be deposited into the Stability Pool.

Reviewer 2: ```suggestion
If this scenario was to arise early into the life of MUSD where a $100M liquidation emptied the Stability Pool, this might leave $109M of BTC in the Stability Pool and insufficient BTC in the Active Pool to honour redemptions for all of the circulating MUSD. In this event the governance process that manages the PCV contract would withdraw the BTC from the Stability Pool and open a loan to borrow MUSD against BTC to get the BTC back into the Active Pool so that redemptions are able to honour requests. The borrowed MUSD would then be deposited into the Stability Pool.
```

Just noting we should probably have an incident response plan for this so we can detail how to do it and who will do it.

##### Rebalancing Small Liquidations

For smaller liquidations the governance process that manages the PCV contract would withdraw BTC from the Stability Pool and swap to MUSD then redeposit into the Stability Pool.

When the protocol's Stability Pool deposit is used to offset liquidations, that results in BTC from the liquidated loans to be in the Stability Pool. The governance process is able to withdraw the BTC to exchange it into MUSD to redeposit into the Stability Pool.

![Rebalancing BTC into MUSD](images/rebalancingBTCintoMUSD.png)

Flow of Funds

1. Governance requests withdrawing BTC from Stability Pool
2. BTC is transferred from Stability Pool to PCV
3. Governance withdraws BTC
4. Governance swaps BTC
5. Governance recieves MUSD
6. Governance deposits MUSD back into PCV
7. PCV transfers MUSD back to Stability Pool

Initially this process would be manually executed via a governance multisig and eventually automated with a smart contract once routing liqudity on Mezo has matured. This is to ensure the Protocol Owned Liquidity does not get raided by MEV while the infrastructure for the new chain matures.

#### Liquidations and the last loan

> The system requires atleast one active loan, if there is only one loan in the system it can not be liquidated. Liquidations require there to be other troves to distribute the debt and collateral to if the Stability Pool is empty.

Reviewer 2: ```suggestion
The system requires at least one active loan, if there is only one loan in the system it can not be liquidated. Liquidations require there to be other troves to distribute the debt and collateral to if the Stability Pool is empty.
```

#### Liquidations and bad debt

> It is anticipated that bad debt would be a low probability event due to the following:

Reviewer 2: Probably worth defining bad debt here.

Someone's full debt can be cancelled in this way. For example, if Carol redeemed $1000 instead of $50, then Alice's debt would be fully paid, and she would be left with $300 worth of collateral. The remaining collateral is sent to the `CollSurplusPool`. Alice can collect it by calling `BorrowerOperations.claimCollateral`.

> The `minNetDebt` value is used to prevent the list of active loans being filled up with nearly empty troves, filling the sorted loans list with lots of nearly empty troves could be used to cause redemption requests to fail by running out of gas. Note that the redemption amount must leave the loan with a debt level greater than the `minNetDebt` value, this is initialised at 1800 MUSD but can be changed through governance, it can be increased or reduced as low as 50 MUSD.

Reviewer 2: It's a little unclear what would happen to existing loans if we increased this value.  Now all the sudden I can't e.g. adjust my trove without putting it above the new minimum.  I guess that's reasonable but it's a feel bad moment for the user.  Perhaps it's worth figuring out how we would handle that (or if we would even do it)?
